lib/libutee/tee_api.c

```diff
@@ -231,6 +234,124 @@ TEE_Result TEE_InvokeTACommand(TEE_TASessionHandle session,
+	if (!tee_api_ocall_session) {
+		res = TEE_OpenTASession(&uuid, TEE_TIMEOUT_INFINITE, 0, NULL,
+					&tee_api_ocall_session, &ret_origin);
+		if (res != TEE_SUCCESS) {
+			if (res == TEE_ERROR_ITEM_NOT_FOUND) {
+				res = TEE_ERROR_NOT_SUPPORTED;
+				ret_origin = TEE_ORIGIN_API;
+			}
+
+			goto exit;
+		}
+	}
+
+	/* Convert the OCALL's parameters into a utee_params structure */
+	ocall_up.types = paramTypes;
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		switch (TEE_PARAM_TYPE_GET(paramTypes, n)) {
+		case TEE_PARAM_TYPE_VALUE_INPUT:
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+			ocall_up.vals[n * 2] = params[n].value.a;
+			ocall_up.vals[n * 2 + 1] = params[n].value.b;
+			break;
+		case TEE_PARAM_TYPE_MEMREF_INPUT:
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+			buffer = params[n].memref.buffer;
+			size = params[n].memref.size;
+			if ((buffer && !size) || (!buffer && size)) {
+				res = TEE_ERROR_BAD_PARAMETERS;
+				ret_origin = TEE_ORIGIN_API;
+				goto exit;
+			}
+			ocall_up.vals[n * 2] = (vaddr_t)buffer;
+			ocall_up.vals[n * 2 + 1] = size;
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* Construct the parameters for the call to the PTA */
+	pta_up.vals[0] = commandID;
+	pta_up.vals[2] = (uintptr_t)&ocall_up;
+	pta_up.vals[3] = sizeof(ocall_up);
+	pta_up.types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INOUT,
+				       TEE_PARAM_TYPE_MEMREF_INOUT,
+				       TEE_PARAM_TYPE_NONE,
+				       TEE_PARAM_TYPE_NONE);
+
+	/* Send the OCALL request */
+	res = utee_invoke_ta_command((uintptr_t)tee_api_ocall_session,
+				     cancellationRequestTimeout,
+				     PTA_OCALL_SEND,
+				     &pta_up,
+				     &ret_origin);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	/* Convert the utee_params structure into the OCALL's parameters */
+	for (n = 0; n < TEE_NUM_PARAMS; n++) {
+		switch (TEE_PARAM_TYPE_GET(paramTypes, n)) {
+		case TEE_PARAM_TYPE_VALUE_OUTPUT:
+		case TEE_PARAM_TYPE_VALUE_INOUT:
+			params[n].value.a = ocall_up.vals[n * 2];
+			params[n].value.b = ocall_up.vals[n * 2 + 1];
+			break;
+		case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+		case TEE_PARAM_TYPE_MEMREF_INOUT:
+			if (ocall_up.vals[n * 2] !=
+			    (vaddr_t)params[n].memref.buffer ||
+			    ocall_up.vals[n * 2 + 1] > params[n].memref.size) {
+				res = TEE_ERROR_BAD_PARAMETERS;
+				ret_origin = TEE_ORIGIN_API;
+				goto exit;
+			}
+			params[n].memref.size = ocall_up.vals[n * 2 + 1];
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* Extract the OCALL return value and error origin */
+	res = (TEE_Result)pta_up.vals[0];
+	ret_origin = (uint32_t)pta_up.vals[1];
+
+exit:
+	/* The PTA is a communications conduit to normal world */
+	if (ret_origin == TEE_ORIGIN_TRUSTED_APP)
+		ret_origin = TEE_ORIGIN_COMMS;
+
+	if (returnOrigin)
+		*returnOrigin = ret_origin;
+
+	if (ret_origin == TEE_ORIGIN_TEE &&
+	    (res != TEE_SUCCESS ||
+	     res != TEE_ERROR_OUT_OF_MEMORY ||
+	     res != TEE_ERROR_TARGET_DEAD))
+		TEE_Panic(res);
```

Comment:
@etienne-lms: This little tidbit is back. I realized that if we panic on things like bad parameters regardless of the error origin, then the few tests we perform above in this function would cause a panic. It seems to me that'd be too harsh.

Do you see any security concerns regarding this more permissive logic (in comparison with `TEE_InvokeTACommand`)?

Reply:
I'll look at that.

Reply:
Maybe we can return early without panicking from anywhere above the `exit` label, except for an error in the return code of `utee_invoke_ta_command`; then, we'd jump to the `exit` label and perhaps decide to panic.